client/client.go

```diff
@@ -230,13 +230,25 @@ func NewClient(cfg *config.Config, consulCatalog consul.CatalogAPI, consulServic
 		return nil, fmt.Errorf("node setup failed: %v", err)
 	}
 
+	fingerprintManager := &FingerprintManager{
+		getConfig: func() *config.Config {
+			c.configLock.Lock()
+			defer c.configLock.Unlock()
+			return c.config
+		},
+		node:       c.config.Node,
+		updateNode: c.updateNodeFromFingerprint,
+		shutdownCh: c.shutdownCh,
+		logger:     c.logger,
+	}
+
 	// Fingerprint the node
-	if err := c.fingerprint(); err != nil {
+	if err := c.fingerprint(fingerprintManager); err != nil {
```

Comment:
I would move both of these into the fingerprint manager. Create a `Run() error` method that does the initial setup and then calls `run`.